linera-core/src/chain_worker/state/mod.rs

```diff
@@ -416,11 +416,53 @@ where
                 .or_default()
                 .insert(target.medium, heights);
         }
+        let heights = BTreeSet::from_iter(
+            heights_by_recipient
+                .iter()
+                .flat_map(|(_, height_map)| height_map.iter().flat_map(|(_, vec)| vec).copied()),
+        );
+        let heights_usize = heights
+            .iter()
+            .copied()
+            .map(usize::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
+        let hashes = self
+            .chain
+            .confirmed_log
+            .multi_get(heights_usize.clone())
+            .await?
+            .into_iter()
+            .zip(heights_usize)
+            .map(|(maybe_hash, height)| {
+                maybe_hash.ok_or_else(|| ViewError::not_found("confirmed log entry", height))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let certificates = self.storage.read_certificates(hashes).await?;
+        let certificates = heights
+            .into_iter()
+            .zip(certificates)
+            .collect::<HashMap<_, _>>();
```

Comment:
Can you extract these to a function(s) please? I think it's obfuscating the purpose of the logic here.